Add support to `◁think▷...◁/think▷` format and DRY the thinking processing logic

allozaur opened the pull request:
- Adds support for another thinking format
- Improves the code architecture for `thinking.ts`

allozaur committed:
feat: Add support to `◁think▷...◁/think▷` format and DRY the thinking processing code

ExtReMLapin commented:
Why and how in hell did we end up getting the the frontend to parse thinking tags ?

The backend returns thinking content inside dedicated field

allozaur commented:
> Why and how in hell did we end up getting the the frontend to parse thinking tags ?
> 
> The backend returns thinking content inside dedicated field

Parsing thinking content on the frontend had been around since the previous version of WebUI. It's necessary as there are cases where we are not getting thinking content directly in the message instead of `reasoning_content`, which of course is supported as you can see below:

Some models do return thinking content in `reasoning_content` and frontend handles this:

https://github.com/ggml-org/llama.cpp/blob/132d673554e65282e92505f4f77401ab971528bb/tools/server/webui/src/lib/services/chat.ts#L298

https://github.com/ggml-org/llama.cpp/blob/132d673554e65282e92505f4f77401ab971528bb/tools/server/webui/src/lib/components/app/chat/ChatMessages/ChatMessage.svelte#L48-L59

https://github.com/ggml-org/llama.cpp/blob/132d673554e65282e92505f4f77401ab971528bb/tools/server/webui/src/lib/stores/chat.svelte.ts#L329-L346

ExtReMLapin commented:
So it's some kind of "Not implemented on backend (cpp code), but faster to implement on frontend" ?

ServeurpersoCom committed:
Merge branch 'pr-16364' into test-think-tags

ServeurpersoCom committed:
fix: capture inline <think> during stream until reasoning_content arrives

- Captured inline <think> segments during streaming, forwarding them to the reasoning UI while keeping the cleaned assistant message stream intact
- Tracked when explicit reasoning_content chunks arrive so inline capture is skipped once the server provides dedicated reasoning updates

ggerganov commented on tools/server/webui/src/lib/utils/thinking.ts:
```diff
@@ -68,31 +53,27 @@ export function parseThinkingContent(content: string): {
-		content.includes('<think>') ||
-		content.includes('[THINK]') ||
+		THINKING_FORMATS.some((format) => content.includes(format.startTag)) ||
 		content.includes('<|channel|>analysis')
```
Thanks for the support. Yes, this is what I figured as well, though my understanding of the chat parsing logic is very rudimentary. Extra eyes on this is appreciated.

ServeurpersoCom committed:
Simplified hasThinkingStart to rely solely on known thinking tag formats, dropping the redundant <|channel|>analysis check now handled upstream

allozaur committed:
Merge remote-tracking branch 'origin/master' into 16158-think-tags

allozaur committed:
Merge remote-tracking branch 'ServeurpersoCom/test-think-tags' into 16158-think-tags

allozaur committed:
chore: Update webui static build

allozaur committed:
chore: Formatting & updated static build

ServeurpersoCom commented:
There's still one tricky edge case that isn't handled: some models expect the \<think> tag to already be opened in the system prompt to start the chain-of-thought, and they were only trained to close it. With SFT done that way, compatibility with other models wasn't really considered because on the very first chunk, how do you know if it's reasoning or the final answer? That means we'd have to hook into /prop / Jinja template again just to propagate extra info, but that feels like a brittle workaround and it doesn't really align with the spirit of OpenAI-Compat.

ExtReMLapin commented:
From memory that's what the thinking 2507 QWEN3 model does. They don't open it, we have to consider it already opened.

ExtReMLapin commented:
Backend can uses the chat template to know what to do 

* Does this model has thinking tags ? only a closing one ?


Exposing the chat template to the client using an endpoint could fix this issue.

But again I still think it would be better to fully implement the logic on the backend and have nothing on the frontend that handles it

ServeurpersoCom commented:
> Backend can uses the chat template to know what to do
> 
>     * Does this model has thinking tags ? only a closing one ?
> 
> 
> Exposing the chat template to the client using an endpoint could fix this issue.
> 
> But again I still think it would be better to fully implement the logic on the backend and have nothing on the frontend that handles it

Right now the handling of reasoning/thinking tags is split between backend and frontend. For GPT-OSS/Harmony the backend already parses <|channel|>analysis and streams it into delta.reasoning_content, so the WebUI just consumes message.reasoning_content.

For other models (Qwen, GLM, etc.) the WebUI still uses legacy checks like content.includes("<think>") or content.includes("[THINK]"), which duplicates logic and makes the frontend fragile.

I think also it would be more appropriate to centralize everything in the backend and I find the idea interesting. I will try an implementation this weekend:

common_chat_parse and helpers detect all formats (<think>, [THINK], <|channel|>analysis, etc.).

The parsed reasoning always goes into message.reasoning_content.

If reasoning_in_content = true, the backend can re-inject it into message.content for legacy clients.

The diff and JSON serialization already handle reasoning_content_delta, so OpenAI-compat output stays consistent.

This way the frontend no longer parses tags, it only reads message.reasoning_content. All models are normalized, the API is consistent, and adding a new model format only requires updating the backend parser once.

allozaur committed:
Merge remote-tracking branch 'ServeurpersoCom/test-think-tags' into 16158-think-tags

allozaur committed:
chore: update webui build output

allozaur commented:
@ggerganov @ngxson i think that we probably could close this PR in favour of #16394

Lemme know what u guys think! :)

allozaur commented:
I think that #16394 is the way to go. Let's continue the reasoning content parsing thread over there. I will close this PR (we can always revisit/reopen if needed).
